### ISSUE TITLE
Send promises never complete?

### ISSUE THREAD

**Issue by christav** [noodlefrenzy/node-amqp10#314]
I've started playing with amqp10 and a service bus queue. I'm having trouble with the `sender.send` method. In specific, while the message does in fact get sent, the promise returned from the `send` method never completes. Am I doing something wrong here?

Running on Windows 10, node 6.9.5. Also tried with node 6.10.2, same behavior.

Sample code:
```javascript
'use strict'

const Promise = require('bluebird')
const { Client: AMQPClient, Policy } = require('amqp10')
const { inspect } = require('util')

const { busName, queueName, sasKeyName, sasAccessKey, amqpUri: uri } = require('./queueConfig')

function formatMessage(count) {
  return {
    body: `This is message #${count + 1}`,
    options: {
      callback: 'sent',
      messageAnnotations: {
        'x-opt-partition-key': 'MySession'
      }
    }
  }
}

let maxMessages = 1000
let start

var client = new AMQPClient(Policy.ServiceBusQueue)
client.connect(uri)
  .then(() => client.createSender(queueName))
  .then(sender => {
    sender.on('errorReceived', tx_err => console.warn(`===> TX ERROR: ${tx_err}`))
    start = process.hrtime()

    let message = formatMessage(0)
    let p = sender.send(message.body, message.options)
      .then(state => console.log(`Send completed with status ${state}`))
    return p

  })
  .then(() => {
    console.log('sending complete')
    // Can't disconnect right now otherwise things blow up since send isn't really done
    //client.disconnect()
    let end = process.hrtime(start)
    let elapsedNs = end[0] * 1e9 + end[1]
    let elapsedSec = elapsedNs / 1e9
    console.log(`Wrote ${maxMessages} in ${elapsedSec}, or ${maxMessages/elapsedSec} messages/second`)
  })
  .catch(err => {
    console.log(`Write failed, error ${err}`)
  })
```

And the console output (with debug turned on):
```
> node .\send-amqp.js
  amqp10:client connecting to: amqps://ReadWrite:<redacted>@nodeperftest.servicebus.windows.net +0ms
  amqp10:transport:tls Connecting to nodeperftest.servicebus.windows.net:5671 via TLS +38ms
  amqp10:connection stateChange: DISCONNECTED => START , reason: connect +14ms
  amqp10:connection Sending Header 414d515003010000 +70ms
  amqp10:connection stateChange: START => IN_SASL , reason: connected +3ms
  amqp10:trace Rx: 414d515003010000 +6ms
  amqp10:sasl Server SASL Version: 414d515003010000 vs 414d515003010000 +3ms
  amqp10:trace Rx: 0000003f02010000005340c03201e02f04b3000000074d53534243425300000005504c41494e00000009414e4f4e594d4f55530000000845585445524e414c +53ms
  amqp10:framing <= @sasl-mechanisms(64) [saslServerMechanisms=[ 'MSSBCBS', 'PLAIN', 'ANONYMOUS', 'EXTERNAL' ]] +4ms
  amqp10:trace raw: [005340c03201e02f04b3000000074d53534243425300000005504c41494e00000009414e4f4e594d4f55530000000845585445524e414c] +5ms
  amqp10:sasl Sending ReadWrite:GvNrdKoxoUvhJ7JnBcxvaUZZ+vnG8qVGvsbztj5uSZs= +1ms
  amqp10:framing => @sasl-init(65) [mechanism='PLAIN' initialResponse=<Buffer 00 52 65 61 64 57 72 69 74 65 00 47 76 4e 72 64 4b 6f 78 6f 55 76 68 4a 37 4a 6e 42 63 78 76 61 55 5a 5a 2b 76 6e 47 38 71 56 47 76 73 62 7a 74 6a 35 ... > hostname=null] +4ms
  amqp10:trace raw: [0000004f02010000005341c04203a305504c41494ea037005265616457726974650047764e72644b6f786f5576684a374a6e4263787661555a5a2b766e47387156477673627a746a3575535a733d40] +2ms
  amqp10:trace Rx: 0000001a02010000005344c00d025000a00857656c636f6d6521 +7ms
  amqp10:framing <= @sasl-outcome(68) [code=0 additionalData=<Buffer 57 65 6c 63 6f 6d 65 21>] +2ms
  amqp10:trace raw: [005344c00d025000a00857656c636f6d6521] +12ms
  amqp10:connection Sending Header 414d515000010000 +1ms
  amqp10:connection stateChange: IN_SASL => HDR_SENT , reason: success +1ms
  amqp10:trace Rx: 414d515000010000 +7ms
  amqp10:connection Server AMQP Version: 414d515000010000 vs 414d515000010000 +10ms
  amqp10:connection stateChange: HDR_SENT => HDR_EXCH , reason: validVersion +0ms
  amqp10:framing => @open(16) [containerId='conn1492533455922' hostname='nodeperftest.servicebus.windows.net' maxFrameSize=4294967295 channelMax=65535 idleTimeout=120000 outgoingLocales='en-US' incomingLocales='en-US' offeredCapabilities=null desiredCapabilities=null properties={}] +7ms
  amqp10:trace raw: [0000006602000000005310c0590aa111636f6e6e31343932353333343535393232a1236e6f646570657266746573742e736572766963656275732e77696e646f77732e6e657470ffffffff60ffff700001d4c0a305656e2d5553a305656e2d55534040c10100] +3ms
  amqp10:connection stateChange: HDR_EXCH => OPEN_SENT , reason: sendOpen +1ms
  amqp10:trace Rx: 0000004702000000005310c03a0aa12433653566393639306230666534646563396561306261363761313366653265345f473233407000010000601387700003a9804040404040 +7ms
  amqp10:framing <= @open(16) [containerId='3e5f9690b0fe4dec9ea0ba67a13fe2e4_G23' hostname=null maxFrameSize=65536 channelMax=4999 idleTimeout=240000 outgoingLocales='en-US' incomingLocales='en-US' offeredCapabilities=null desiredCapabilities=null properties={}] +1ms
  amqp10:trace raw: [005310c03a0aa12433653566393639306230666534646563396561306261363761313366653265345f473233407000010000601387700003a9804040404040] +1ms
  amqp10:connection stateChange: OPEN_SENT => OPENED , reason: openReceived +1ms
  amqp10:connection Connected with params { maxFrameSize= 65536 , channelMax= 4999 , idleTimeout= 120000 , remoteIdleTimeout= 240000  } +1ms
  amqp10:connection Setting heartbeat check timeout to 30000 +11ms
  amqp10:client connected +1ms
  amqp10:session stateChange(undefined): UNMAPPED => BEGIN_SENT , reason: sendBegin +2ms
  amqp10:framing => @begin(17) [remoteChannel=null nextOutgoingId=1 incomingWindow=100 outgoingWindow=100 handleMax=4294967295 offeredCapabilities=null desiredCapabilities=null properties={}] +1ms
  amqp10:trace raw: [0000001f02000001005311c012084052015264526470ffffffff4040c10100] +1ms
  amqp10:trace Rx: 0000002102000000005311c0140860000152017000001388526452ff4040c10100 +22ms
  amqp10:framing <= @begin(17) [remoteChannel=1 nextOutgoingId=1 incomingWindow=5000 outgoingWindow=100 handleMax=255 offeredCapabilities=null desiredCapabilities=null properties={}] +2ms
  amqp10:trace raw: [005311c0140860000152017000001388526452ff4040c10100] +1ms
  amqp10:session stateChange(undefined): BEGIN_SENT => MAPPED , reason: beginReceived +1ms
  amqp10:session On BEGIN_RCVD, setting params to (1,5000,100,255) +0ms
  amqp10:client mapped +1ms
  amqp10:link stateChange(undefined:0): DETACHED => ATTACHING , reason: sendAttach +5ms
  amqp10:framing => @attach(18) [name='nodeperftest_7ecdcdb7-e923-47ce-afbd-f9f55e28f678' handle=0 role=false sndSettleMode=1 rcvSettleMode=0 source=@source(40) [address='localhost' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} distributionMode=null filter={} defaultOutcome=null outcomes=null capabilities=null] target=@target(41) [address='nodeperftest' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} capabilities=null] unsettled={} incompleteUnsettled=false initialDeliveryCount=1 maxMessageSize=10000 offeredCapabilities=null desiredCapabilities=null properties={}] +3ms
  amqp10:trace raw: [000000ae02000001005312c0a10ea1316e6f646570657266746573745f37656364636462372d653932332d343763652d616662642d663966353565323866363738434250015000005328c0260ba1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100404040005329c02307a10c6e6f6465706572667465737443a30b73657373696f6e2d656e644342c1010040c101004252018000000000000027104040c10100] +3ms
  amqp10:trace Rx: 000000ec02000000005312c0df0ea1316e6f646570657266746573745f37656364636462372d653932332d343763652d616662642d663966353565323866363738434150015000005328c0260ba1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100404040005329c02307a10c6e6f6465706572667465737443a30b73657373696f6e2d656e644342c10100404040408000000000000027104040c14202a319636f6d2e6d6963726f736f66743a747261636b696e672d6964a12433653566393639306230666534646563396561306261363761313366653265345f473233 +8ms
  amqp10:framing <= @attach(18) [name='nodeperftest_7ecdcdb7-e923-47ce-afbd-f9f55e28f678' handle=0 role=true sndSettleMode=1 rcvSettleMode=0 source=@source(40) [address='localhost' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} distributionMode=null filter={} defaultOutcome=null outcomes=null capabilities=null] target=@target(41) [address='nodeperftest' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} capabilities=null] unsettled={} incompleteUnsettled=false initialDeliveryCount=null maxMessageSize=10000 offeredCapabilities=null desiredCapabilities=null properties={ 'com.microsoft:tracking-id': '3e5f9690b0fe4dec9ea0ba67a13fe2e4_G23' }] +6ms
  amqp10:trace raw: [005312c0df0ea1316e6f646570657266746573745f37656364636462372d653932332d343763652d616662642d663966353565323866363738434150015000005328c0260ba1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100404040005329c02307a10c6e6f6465706572667465737443a30b73657373696f6e2d656e644342c10100404040408000000000000027104040c14202a319636f6d2e6d6963726f736f66743a747261636b696e672d6964a12433653566393639306230666534646563396561306261363761313366653265345f473233] +2ms
  amqp10:link stateChange(nodeperftest_7ecdcdb7-e923-47ce-afbd-f9f55e28f678:0): ATTACHING => ATTACHED , reason: attachReceived +4ms
  amqp10:link nodeperftest_7ecdcdb7-e923-47ce-afbd-f9f55e28f678: attached CH=[1=>0], Handle=[0=>0] +1ms
  amqp10:link:sender canSend(0,5000) = false +3ms
Promise {
  _bitField: 0,
  _fulfillmentHandler0: undefined,
  _rejectionHandler0: undefined,
  _promise0: undefined,
  _receiver0: undefined }
  amqp10:trace Rx: 0000002402000000005313c0170b52017000001388520152644343700000012c43404240 +40ms
  amqp10:framing <= @flow(19) [nextIncomingId=1 incomingWindow=5000 nextOutgoingId=1 outgoingWindow=100 handle=0 deliveryCount=0 linkCredit=300 available=0 drain=false echo=false properties={}] +0ms
  amqp10:trace raw: [005313c0170b52017000001388520152644343700000012c43404240] +2ms
  amqp10:session New Incoming Window (known id): 5000 = 1 + 5000 - 1 +1ms
  amqp10:link:sender setting credits (300,5000) +0ms
  amqp10:link:sender canSend(300,5000) = true +4ms
  amqp10:link:sender sending:  <Buffer 54 68 69 73 20 69 73 20 6d 65 73 73 61 67 65 20 23 31> +0ms
  amqp10:framing => @transfer(20) [handle=0 deliveryId=1 deliveryTag=<Buffer 31> messageFormat=0 settled=true more=false rcvSettleMode=0 state=null resume=false aborted=false batchable=false] +5ms
  amqp10:trace raw: [0000006002000001005314c0100b435201a00131434142500040424242005372c12702a313782d6f70742d706172746974696f6e2d6b6579a10f506572665465737453657373696f6e005375a01254686973206973206d657373616765202331] +1ms
  amqp10:trace Rx: 0000000802000000 +2m
  amqp10:framing <= (EMPTY FRAME) +2ms
  amqp10:trace raw: [0000000802000000] +1ms
  amqp10:framing => (EMPTY FRAME) +45s
  amqp10:trace raw: [0000000802000000] +2ms
  amqp10:trace Rx: 0000000802000000 +1m
  amqp10:framing <= (EMPTY FRAME) +3ms
  amqp10:trace raw: [0000000802000000] +0ms
  amqp10:framing => (EMPTY FRAME) +60s
  amqp10:trace raw: [0000000802000000] +1ms
  amqp10:trace Rx: 0000000802000000 +45s
  amqp10:framing <= (EMPTY FRAME) +1ms
  amqp10:trace raw: [0000000802000000] +0ms
  amqp10:framing => (EMPTY FRAME) +1m
  amqp10:trace raw: [0000000802000000] +55ms
  amqp10:trace Rx: 0000000802000000 +30s
  amqp10:framing <= (EMPTY FRAME) +1ms
  amqp10:trace raw: [0000000802000000] +1ms
dev:\sbperf
> node .\send-amqp.js
  amqp10:client connecting to: amqps://ReadWrite:GvNrdKoxoUvhJ7JnBcxvaUZZ%2BvnG8qVGvsbztj5uSZs%3D@nodeperftest.servicebus.windows.net +0ms
  amqp10:transport:tls Connecting to nodeperftest.servicebus.windows.net:5671 via TLS +40ms
  amqp10:connection stateChange: DISCONNECTED => START , reason: connect +31ms
  amqp10:connection Sending Header 414d515003010000 +362ms
  amqp10:connection stateChange: START => IN_SASL , reason: connected +2ms
  amqp10:trace Rx: 414d515003010000 +11ms
  amqp10:sasl Server SASL Version: 414d515003010000 vs 414d515003010000 +4ms
  amqp10:trace Rx: 0000003f02010000005340c03201e02f04b3000000074d53534243425300000005504c41494e00000009414e4f4e594d4f55530000000845585445524e414c +50ms
  amqp10:framing <= @sasl-mechanisms(64) [saslServerMechanisms=[ 'MSSBCBS', 'PLAIN', 'ANONYMOUS', 'EXTERNAL' ]] +6ms
  amqp10:trace raw: [005340c03201e02f04b3000000074d53534243425300000005504c41494e00000009414e4f4e594d4f55530000000845585445524e414c] +10ms
  amqp10:sasl Sending ReadWrite:GvNrdKoxoUvhJ7JnBcxvaUZZ+vnG8qVGvsbztj5uSZs= +3ms
  amqp10:framing => @sasl-init(65) [mechanism='PLAIN' initialResponse=<Buffer 00 52 65 61 64 57 72 69 74 65 00 47 76 4e 72 64 4b 6f 78 6f 55 76 68 4a 37 4a 6e 42 63 78 76 61 55 5a 5a 2b 76 6e 47 38 71 56 47 76 73 62 7a 74 6a 35 ... > hostname=null] +4ms
  amqp10:trace raw: [0000004f02010000005341c04203a305504c41494ea037005265616457726974650047764e72644b6f786f5576684a374a6e4263787661555a5a2b766e47387156477673627a746a3575535a733d40] +21ms
  amqp10:trace Rx: 0000001a02010000005344c00d025000a00857656c636f6d6521 +12ms
  amqp10:framing <= @sasl-outcome(68) [code=0 additionalData=<Buffer 57 65 6c 63 6f 6d 65 21>] +6ms
  amqp10:trace raw: [005344c00d025000a00857656c636f6d6521] +13ms
  amqp10:connection Sending Header 414d515000010000 +2ms
  amqp10:connection stateChange: IN_SASL => HDR_SENT , reason: success +1ms
  amqp10:trace Rx: 414d515000010000 +6ms
  amqp10:connection Server AMQP Version: 414d515000010000 vs 414d515000010000 +5ms
  amqp10:connection stateChange: HDR_SENT => HDR_EXCH , reason: validVersion +1ms
  amqp10:framing => @open(16) [containerId='conn1492533899568' hostname='nodeperftest.servicebus.windows.net' maxFrameSize=4294967295 channelMax=65535 idleTimeout=120000 outgoingLocales='en-US' incomingLocales='en-US' offeredCapabilities=null desiredCapabilities=null properties={}] +10ms
  amqp10:trace raw: [0000006602000000005310c0590aa111636f6e6e31343932353333383939353638a1236e6f646570657266746573742e736572766963656275732e77696e646f77732e6e657470ffffffff60ffff700001d4c0a305656e2d5553a305656e2d55534040c10100] +2ms
  amqp10:connection stateChange: HDR_EXCH => OPEN_SENT , reason: sendOpen +2ms
  amqp10:trace Rx: 0000004702000000005310c03a0aa12465343731343362306266613534666432386262303265316138303063306634385f473132407000010000601387700003a9804040404040 +7ms
  amqp10:framing <= @open(16) [containerId='e47143b0bfa54fd28bb02e1a800c0f48_G12' hostname=null maxFrameSize=65536 channelMax=4999 idleTimeout=240000 outgoingLocales='en-US' incomingLocales='en-US' offeredCapabilities=null desiredCapabilities=null properties={}] +1ms
  amqp10:trace raw: [005310c03a0aa12465343731343362306266613534666432386262303265316138303063306634385f473132407000010000601387700003a9804040404040] +1ms
  amqp10:connection stateChange: OPEN_SENT => OPENED , reason: openReceived +41ms
  amqp10:connection Connected with params { maxFrameSize= 65536 , channelMax= 4999 , idleTimeout= 120000 , remoteIdleTimeout= 240000  } +1ms
  amqp10:connection Setting heartbeat check timeout to 30000 +2ms
  amqp10:client connected +2ms
  amqp10:session stateChange(undefined): UNMAPPED => BEGIN_SENT , reason: sendBegin +2ms
  amqp10:framing => @begin(17) [remoteChannel=null nextOutgoingId=1 incomingWindow=100 outgoingWindow=100 handleMax=4294967295 offeredCapabilities=null desiredCapabilities=null properties={}] +3ms
  amqp10:trace raw: [0000001f02000001005311c012084052015264526470ffffffff4040c10100] +0ms
  amqp10:trace Rx: 0000002102000000005311c0140860000152017000001388526452ff4040c10100 +13ms
  amqp10:framing <= @begin(17) [remoteChannel=1 nextOutgoingId=1 incomingWindow=5000 outgoingWindow=100 handleMax=255 offeredCapabilities=null desiredCapabilities=null properties={}] +1ms
  amqp10:trace raw: [005311c0140860000152017000001388526452ff4040c10100] +1ms
  amqp10:session stateChange(undefined): BEGIN_SENT => MAPPED , reason: beginReceived +1ms
  amqp10:session On BEGIN_RCVD, setting params to (1,5000,100,255) +1ms
  amqp10:client mapped +1ms
  amqp10:link stateChange(undefined:0): DETACHED => ATTACHING , reason: sendAttach +10ms
  amqp10:framing => @attach(18) [name='nodeperftest_db9a3bfe-892f-4990-8807-93b82ffc9357' handle=0 role=false sndSettleMode=1 rcvSettleMode=0 source=@source(40) [address='localhost' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} distributionMode=null filter={} defaultOutcome=null outcomes=null capabilities=null] target=@target(41) [address='nodeperftest' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} capabilities=null] unsettled={} incompleteUnsettled=false initialDeliveryCount=1 maxMessageSize=10000 offeredCapabilities=null desiredCapabilities=null properties={}] +4ms
  amqp10:trace raw: [000000ae02000001005312c0a10ea1316e6f646570657266746573745f64623961336266652d383932662d343939302d383830372d393362383266666339333537434250015000005328c0260ba1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100404040005329c02307a10c6e6f6465706572667465737443a30b73657373696f6e2d656e644342c1010040c101004252018000000000000027104040c10100] +1ms
  amqp10:trace Rx: 000000ec02000000005312c0df0ea1316e6f646570657266746573745f64623961336266652d383932662d343939302d383830372d393362383266666339333537434150015000005328c0260ba1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100404040005329c02307a10c6e6f6465706572667465737443a30b73657373696f6e2d656e644342c10100404040408000000000000027104040c14202a319636f6d2e6d6963726f736f66743a747261636b696e672d6964a12465343731343362306266613534666432386262303265316138303063306634385f473132 +9ms
  amqp10:framing <= @attach(18) [name='nodeperftest_db9a3bfe-892f-4990-8807-93b82ffc9357' handle=0 role=true sndSettleMode=1 rcvSettleMode=0 source=@source(40) [address='localhost' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} distributionMode=null filter={} defaultOutcome=null outcomes=null capabilities=null] target=@target(41) [address='nodeperftest' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} capabilities=null] unsettled={} incompleteUnsettled=false initialDeliveryCount=null maxMessageSize=10000 offeredCapabilities=null desiredCapabilities=null properties={ 'com.microsoft:tracking-id': 'e47143b0bfa54fd28bb02e1a800c0f48_G12' }] +12ms
  amqp10:trace raw: [005312c0df0ea1316e6f646570657266746573745f64623961336266652d383932662d343939302d383830372d393362383266666339333537434150015000005328c0260ba1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100404040005329c02307a10c6e6f6465706572667465737443a30b73657373696f6e2d656e644342c10100404040408000000000000027104040c14202a319636f6d2e6d6963726f736f66743a747261636b696e672d6964a12465343731343362306266613534666432386262303265316138303063306634385f473132] +4ms
  amqp10:link stateChange(nodeperftest_db9a3bfe-892f-4990-8807-93b82ffc9357:0): ATTACHING => ATTACHED , reason: attachReceived +2ms
  amqp10:link nodeperftest_db9a3bfe-892f-4990-8807-93b82ffc9357: attached CH=[1=>0], Handle=[0=>0] +1ms
  amqp10:link:sender canSend(0,5000) = false +5ms
  amqp10:trace Rx: 0000002402000000005313c0170b52017000001388520152644343700000012c43404240 +31ms
  amqp10:framing <= @flow(19) [nextIncomingId=1 incomingWindow=5000 nextOutgoingId=1 outgoingWindow=100 handle=0 deliveryCount=0 linkCredit=300 available=0 drain=false echo=false properties={}] +4ms
  amqp10:trace raw: [005313c0170b52017000001388520152644343700000012c43404240] +1ms
  amqp10:session New Incoming Window (known id): 5000 = 1 + 5000 - 1 +0ms
  amqp10:link:sender setting credits (300,5000) +2ms
  amqp10:link:sender canSend(300,5000) = true +39ms
  amqp10:link:sender sending:  <Buffer 54 68 69 73 20 69 73 20 6d 65 73 73 61 67 65 20 23 31> +1ms
  amqp10:framing => @transfer(20) [handle=0 deliveryId=1 deliveryTag=<Buffer 31> messageFormat=0 settled=true more=false rcvSettleMode=0 state=null resume=false aborted=false batchable=false] +5ms
  amqp10:trace raw: [0000005a02000001005314c0100b435201a00131434142500040424242005372c12102a313782d6f70742d706172746974696f6e2d6b6579a1094d7953657373696f6e005375a01254686973206973206d657373616765202331] +38ms
  amqp10:trace Rx: 0000000802000000 +2m
  amqp10:framing <= (EMPTY FRAME) +2ms
  amqp10:trace raw: [0000000802000000] +2ms
  amqp10:framing => (EMPTY FRAME) +45s
  amqp10:trace raw: [0000000802000000] +2ms
```

At this point the process is stopped - the promise never completes, we never see the 'sending complete' message, and things just sit.

Any ideas?

**Comment by mbroadst** [comment]
@christav for some reason it looks like you're not receiving disposition from the remote side, what version of the module are you using?

**Comment by christav** [comment]
@mbroadst It's 3.5.2

The other side is an Azure Service bus queue hosted in the West US 2 region if that helps.

**Comment by mbroadst** [comment]
@christav seems like there might be something up with your queue's configuration (I'm not exactly in a position to help with that, but it looks like you might be). As you can see, we have integration tests running against SB and EH, and those are definitely not failing to resolve.  I'll try to take a look at your debug logs a little more closely when I get home tonight.

As an aside: I see you're doing some performance testing. I was able to get maximal throughput by adding the following [callback](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/policies/policy.js#L135) method.

**Comment by christav** [comment]
Thanks for the suggestion. I just created the queue in the portal with the default settings - anything you can thing of that I should tweak?

The perf testing part is about comparing it to the official Azure SDK for service bus. It's already an order of magnitude faster, so I'm not too worried about squeezing out another couple of percent at the moment. 😄

**Comment by mbroadst** [comment]
@christav I'm on the move right now, but I would reach out to @noodlefrenzy, @pierreca or @princjef, all on your side of the fence

**Comment by princjef** [comment]
I started seeing this too when I moved to 3.5.1.

The problem is with this line:

```
amqp10:framing => @attach(18) [name='nodeperftest_7ecdcdb7-e923-47ce-afbd-f9f55e28f678' handle=0 role=false sndSettleMode=1 rcvSettleMode=0 source=@source(40) [address='localhost' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} distributionMode=null filter={} defaultOutcome=null outcomes=null capabilities=null] target=@target(41) [address='nodeperftest' durable=0 expiryPolicy='session-end' timeout=0 dynamic=false dynamicNodeProperties={} capabilities=null] unsettled={} incompleteUnsettled=false initialDeliveryCount=1 maxMessageSize=10000 offeredCapabilities=null desiredCapabilities=null properties={}] +3ms
```

Your sender is setting the `sndSettleMode` to `1` (settled), which indicates that all messages are already settled when they are received by Service Bus. Based on my understanding of the AMQP spec, you should not receive a disposition back, which appears to be causing the Promise to not resolve.

This is just showing up now because it looks like Service Bus ignored the previous `senderSettleMode` that was sent and would default to `mixed`, which will give you the dispositions you need. Version 3.5.1 fixed the name, causing [the setting in the Service Bus Policy](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/policies/service_bus_policy.js#L9) to be used.

I think there are two parts to this:

 1. **Dispositions not sending** - @christav you can get these by overriding the ServiceBusPolicy with one of the other sender settle modes (I use `mixed`). see [the README](https://github.com/noodlefrenzy/node-amqp10#usage) for instructions on customizing policies

 2. **Promises not resolving when there are no dispositions** - this seems like a bug when using the `settled` sender mode. @mbroadst I'm assuming the promise should still resolve when the sender is in `settled` mode, even with no dispositions sent

**Comment by christav** [comment]
@princjef Thanks, setting the sender settling mode to mixed fixed the promise resolving issue. However, there is a pretty significant perf cost there - from what I could tell sending 1000 messages using the 'settled' default mode takes a couple of seconds (can't reliably time it of course) while with mixed I get about 28 messages per second, so 30-ish seconds total.

I would also strongly suggest that sender.send always return a promise, and if sender mode is settled then that promise should just be pre-resolved.

Also the docs say that sender.send doesn't return a promise in settled mode, but it actually does. So something's not lining up.

Thanks for the help, I'm unblocked for now, but I think there's a legit bug here.

**Comment by mbroadst** [comment]
@princjef thanks for looking into this, you're totally on point here. There is some initial confusion in our policies because we have e.g. `sndSettleMode` and "callback policies".   As for the policy issue, that's spot on as well, I wonder if one of you could verify what the default behavior is? We can have the policy default to that now, or I can just default it to `mixed`.   I'm sort of surprised that our integration tests are working without any issue after this change 🤔.

**Comment by geoffreysamper** [comment]
Hi, I'm having the same issue here when I upgraded to the lasted version

**Comment by mbroadst** [comment]
Okay this should be fixed in master in https://github.com/noodlefrenzy/node-amqp10/commit/be48605645c1e5e472829f2179eb81205a6c2fb0.  I'll wait for you guys to get back to me about what the default should be for ServiceBus/EventHub.

**Comment by princjef** [comment]
@mbroadst sorry for not getting back sooner on this. I think that `mixed` is a reasonable default, especially since it's the default in the AMQP spec and the one that was effectively used up until now. I don't see any special reason why it should default to something different, and people can still decide to set things up in a different way if they want.

I see `settled` as more of a performance/tuning option people can use if they want.

**Comment by princjef** [comment]
Side note to @christav: if you're trying to pump a lot of messages into Service Bus, there's no need to execute them all in sequence. I routinely dump messages into Service Bus with concurrency greater than 100 with no problems. The benefit of getting the disposition back is that you know whether or not Service Bus actually accepted each message

**Comment by mbroadst** [comment]
@christav could you possibly test your code against the `master` branch here and verify that the fixes work for you? I'd prefer to not make multiple patch releases around this issue.

**Comment by christav** [comment]
@mbroadst Sure, I should be able to do that later today.

**Comment by christav** [comment]
@mbroadst Sorry for the long delay, you know how these things go I'm sure. :-) I did try against the master branch, and the send code does in fact work with the default Policy.ServiceBusQueue policy now without having to set the sender settling mode manually.

I also tried setting the sndSettleMode to settled manually, and the promises did complete and perf was a lot better than the default. So I think as far as I am concerned the code currently in master does fix this bug.
